### PR TITLE
Some performance improvements

### PR DESCRIPTION
**Pull request by weiznich**
* Replace the IndexMap in the serialized object with a plain
  `Vec<(String, Value)>` because linear search is faster for few
  elements

* Some general tweaks to skip some allocations

Addresses #199

### CODE REVIEW

**Comment by LegNeato** [comment]
@theduke

**Comment by weiznich** [comment]
The ci fails because something on the master branch is brocken for the `juniper_iron` test case.

File: juniper/src/value.rs [hunk 1]
```diff
@@ -22,7 +21,115 @@ pub enum Value {
...
+        for<'a> &'a str: PartialEq<K>,
+    {
+        self.key_value_list
+            .iter()
+            .any(|(key, _)| (key as &str) == f)
+    }
+
+
+    /// Get a iterator over all field value pairs
+    pub fn iter(&self) -> impl Iterator<Item = &(String, Value)> {
```
Comment (LegNeato) [inline juniper/src/value.rs]: Hmmm, looks like this uses impl trait, which didn't stabilize until [1.26](https://blog.rust-lang.org/2018/05/10/Rust-1.26.html). Supporting impl trait would mean we would have to bump our min rust version.

**Comment by theduke** [comment]
Regarding versions, I'd like to stick to the `stable + last 2 versions` rule we have right now, so impl trait would have to wait for 1.28.

Also, we'd really need some solid benchmarks comparing the two implementations.

This might be a good oppurtunity to set up a nice [criterion](https://github.com/japaric/criterion.rs) based benchmark suite.
It also might be an opportunity to contribute to/ improve the indexmap crate.

**Comment by theduke** [comment]
Either way, as a first step,  it might be a cool idea to export a wrapper type from a submodule so we can easily change the implementation without a lot of hassle.

**Comment by weiznich** [comment]
(I'm back from vacation, so here are some answers.)

> Regarding versions, I'd like to stick to the stable + last 2 versions rule we have right now, so impl trait would have to wait for 1.28.

It should be possible to do this without `impl trait`. I will fix the code in the next few days.

> Also, we'd really need some solid benchmarks comparing the two implementations.

I've developed this change as part of optimizing wundergraph. For this I used the [following benchmark](https://github.com/hasura/graphql-backend-benchmarks). See the wundergraph repo (wundergraph-bench) for commits [without](https://github.com/weiznich/wundergraph/commit/174e3760545fffa615402a14644a73df59dd7f32) and [with](https://github.com/weiznich/wundergraph/commit/6f44e5421d611e7f33a3ef1664754f11ff6a063d) this change. The relevant graphql query is this one:
```graphql
query tracks_media_all {
  Tracks {
    id
    name
    media_type_id {
      name
    }
  }
}
```

Benchmark results without this change:
41662707-36d0ecda-7491-11e8-89d1-9ac1af0bd5b9

Benchmark results with this change:
signal-2018-06-24-202927

> This might be a good oppurtunity to set up a nice criterion based benchmark suite.

My benchmark is unfortunately something that involves a quite complex interplay of several components so it seems not to be appropriated to be used as benchmark for juniper alone. 

> It also might be an opportunity to contribute to/ improve the indexmap crate.

This is not something that could be improved in the indexmap crate.
Basically this PR changes 3 things:

1. It introduces a wrapper type for the object enum kind. This is a purely cosmetic change to give us room to change the internal presentation later.
2. It forces every construction of that wrapper type to preallocate a fixed number of elements. This eliminates the need of continuously reallocating space for more elements as soon as we start to push more field value pairs into the object. (We know how many fields we need in every place so we could just allocate the right number of slots). This contributes ~50% of the performance improvements.
3. It replaces the use of `IndexMap<String, Value>` with a plain `Vec<(String, Value)>`. On the first look this change is making things worse because now we need to to a linear search instead of a hash lookup if we want to know if some field exists. But in my observation for almost all cases that I've seen there are only a few fields (< ~50) so a linear search is faster than doing the hash lookup. This contributes the other ~50%.

**Comment by dvic** [comment]
What is the status of this PR? I believe the `impl trait` has been removed from this PR, right? Are there any remaining issues/todo's before this can be merged?

**Comment by weiznich** [comment]
The usage of `impl Trait` has been already removed in 3b3d0411b8d2c07db7647fb7cfa8fa98298ea78d.
This pr is missing a final review/decision from the juniper team.

**Comment by LegNeato** [comment]
Ah, didn't see this was updated, thanks for the reminder! I'm merging this as-is as I think the change is good and straightforward. I'm slightly concerned about:

1. This is basically implementing a vector-backed arena. It might be better to use one from crates.io, as there are no shortage of them with various perf and mem tradeoffs. Or better yet, have a pluggable interface as @theduke mentioned.

2. Down the road it might not be obvious why we need the arena. I'm worried without some sort of validation or perf test we could regress.

In any case, I don't think those are blockers for landing this.

**Pull request merged**